### ISSUE TITLE
Add useMemo(dependencies vararg, callback) and useCallback(dependencies vararg, callback)

### ISSUE THREAD

**Issue by vlsi** [JetBrains/kotlin-wrappers#304]
```kotlin
fun<T> useMemo(vararg dependencies: dynamic, callback: () -> T) =
    useMemo(callback, dependencies)

fun<T : Function<*>> useCallback(vararg dependencies: dynamic, callback: () -> T) =
    useCallback(callback, dependencies)
```

Sample:
```kotlin
val before = useMemo({ "Hello, ${props.greeting}" }, arrayOf(props.greeting))

val after = useMemo(props.greeting) { "Hello, ${props.greeting}" }
```

**Comment by Leonya** [comment]
How is it different from existing useMemo and useCallback declarations (except for parameter order which, in your case, doesn't match the one in React)

**Comment by vlsi** [comment]
Making the lambda the last argument helps readability a lot, especially for multiline callbacks.

https://github.com/night-crawler/quiz-frontend/blob/fbe4ba2e554c595bda6405ebe2d77e02a7293404/src/main/kotlin/fm/force/ui/hook/UseForceUpdate.kt#L9-L13

```kotlin
    return useCallback(
        {
            dispatch(Object.create(null))
        },
        arrayOf(dispatch)
    )
```

vs

```kotlin
    return useCallback(dispatch) {
        dispatch(Object.create(null))
    }
```

**Comment by Leonya** [comment]
I’m sorry, we are going to follow React here.

**Comment by turansky** [comment]
4 existed effect methods [do not respect React order](https://github.com/JetBrains/kotlin-wrappers/blob/master/kotlin-react/src/main/kotlin/react/hooks.kt#L81) to be more 'Kotlin'

**Comment by turansky** [comment]
@Leonya `useEffect` implementation don't respect React order. Isn't it?

**Comment by Leonya** [comment]
Thanks, I'll check.

**Comment by Skolotsky** [comment]
I agree that Kotlin-like API is more useful. We can keep both of them. Something like 
```kotlin
fun <T> useMemo(vararg args: dynamic, callback: () -> T): T {
    return useMemo(callback, args)
}
```

```kotlin
fun <T> useEffect(vararg args: dynamic, effect: () -> T) {
    val rawEffect = {
        effect()
        undefined
    }
    rawUseEffect(rawEffect, args)
}
```

**Comment by vlsi** [comment]
Just in case, `vararg args: dynamic` needs more investigation, especially for `useEffect`.

The problem is `useEffect(effect: () -> T)` in React means "execute effect on each render".
However, `useEffect(effect: () -> T,  arrayOf())` means "execute only once".

In other words, it is required to distinguish "absence of dependencies" (==execute always) and "empty dependencies" (==execute once)

**Comment by Skolotsky** [comment]
Maybe we can use some wording here like "useEffectAlways" for special case.

**Comment by vlsi** [comment]
>Maybe we can use some wording here like "useEffectAlways" for special case.

useEffect, useEffectWithCleanup, useEffectAlways, useEffectAlwaysWithCleanup is not that bad, but it might multiply quite soon.

However, "always" is not needed for `useMemo` / `useCallback`.

**Comment by Leonya** [comment]
[test](test.com)

**Comment by turansky** [comment]
Fixes:
`useMemo` - #391
`useCallback` - #395

**Comment by vlsi** [comment]
Thanks @turansky , this is indeed fixed by #391 and #395
